### PR TITLE
EOS-23136: Upgrade Framework changes

### PR DESCRIPTION
**Pull request by vimalraghwani20**
# Problem Statement
Implementation of upgrade interface for openldap mini provisioner

# Design
Implement Pre upgrade interface
Implement Post upgrade interface
Upgrade py-utils rpm

# Coding 
-  Coding conventions are followed and code is consistent [Y/N]: 
-  Confirm All CODACY errors are resolved [Y/N]: 

# Testing 
[OpenldapProvisioning.log](https://github.com/Seagate/cortx-utils/files/6892228/OpenldapProvisioning.log)

# Review Checklist 
  Before posting the PR please ensure
- [ ] PR is self reviewed
- [ ] Is there a change in filename/package/module or signature [Y/N]: 
- [ ] If yes for above point, Is a notification sent to all other cortx components [Y/N]
- [ ] Jira is updated
- [ ] Check if the description is clear and explained. 
- [ ] Check Acceptance Criterion is defined. 
- [ ] All the tests performed should be mentioned before Resolving a JIRA. 
- [ ] Verification needs to be done before marked as Closed/Verified 

# Documentation
- [ ] Changes done to WIKI / Confluence page

### CODE REVIEW

File: py-utils/src/utils/setup/openldap/openldap.py [hunk 1]
```diff
@@ -47,14 +47,20 @@ class Openldap:
...
+    machine_id = None
+    cluster_id = None
+    cluster_id_key = None
 
     def __init__(self, conf_url):
+        Conf.load(self.prov, f'yaml://{self._prov_conf_file}')
         if not os.path.isfile(self._preqs_conf_file):
             raise OpenldapSetupError({"message":"pre-requisite json file \
                 not found"})
+        if conf_url is None:
```
Comment (shahidSTX) [inline py-utils/src/utils/setup/openldap/openldap.py:60]: For reset and cleanup, the requirement to have a `conf_url` will be removed. So please remove this check or add it only for phases : post_install, prepare, config, init

File: py-utils/src/utils/setup/openldap/openldap.py [hunk 2]
```diff
@@ -268,3 +278,25 @@ def cleanup(self):
         CleanupCmd(self.url).process()
         Log.debug("%s - Successful" % phase_name)
         return 0
+
+    def preupgrade(self):
+        """Perform pre upgrade action."""
+        phase_name = "preupgrade"
+        Log.debug("%s - Starting\n" % phase_name)
```
Comment (shahidSTX) [inline py-utils/src/utils/setup/openldap/openldap.py:285]: Remove '\n'

File: py-utils/src/utils/setup/openldap/openldap.py [hunk 3]
```diff
@@ -268,3 +278,25 @@ def cleanup(self):
...
+        self.validate(phase_name)
+        self._keys_validate(phase_name)
+        from preupgradecmd import PreUpgradeCmd
```
Comment (shahidSTX) [inline py-utils/src/utils/setup/openldap/openldap.py:288]: Move all imports to the top of this file, in one place.

File: py-utils/src/utils/setup/openldap/preupgradecmd.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+from pathlib import Path
+from cortx.utils.log import Log
+from setupcmd import SetupCmd, OpenldapPROVError
+
+class PreUpgradeCmd(SetupCmd):
+
+  """Pre Upgrade Setup Cmd."""
+
+  name = "preupgrade"
+  utils_tmp_dir = "/opt/seagate/cortx/utils/tmp"
```
Comment (shalakadharap) [inline py-utils/src/utils/setup/openldap/preupgradecmd.py:33]: Fetch it from Conf.get(index= 'util_config_file_index', key='install_path')
then append '/cortx/utils/tmp to it

File: py-utils/src/utils/setup/openldap/openldap_setup.py [hunk 5]
```diff
@@ -47,7 +47,7 @@ def usage(prog: str):
         sys.stderr.write(
             f"usage: {prog} [-h] <cmd> --config <url> <args>...\n"
             f"where:\n"
-            f"cmd   post_install, config, init, reset, test\n"
+            f"cmd   post_install, config, init, reset, test, preupgrade, postupgrade\n"
```
Comment (shahidSTX) [inline py-utils/src/utils/setup/openldap/openldap_setup.py:50]: 'cleanup' phase reference is missing after 'test' which should have been included in separate PR, please add that here if possible.

File: py-utils/src/utils/setup/openldap/openldapsetup_prereqs.json [hunk 6]
```diff
@@ -33,5 +33,17 @@
     "rpms": ["openldap-servers", "openldap-clients", "openldap-devel"],
     "pip3s": ["cryptography", "python-ldap"],
     "services": ["slapd"]
+  },
+  "preupgrade" : {
+    "rpms": ["openldap-servers", "openldap-clients", "openldap-devel"],
+    "pip3s": ["cryptography", "python-ldap"],
+    "files": ["/opt/seagate/cortx/utils/conf/openldap_config.yaml",
```
Comment (shahidSTX) [inline py-utils/src/utils/setup/openldap/openldapsetup_prereqs.json:40]: Is there a package for validating files? If yes, please add that check in validation. If no, please remove this from here.

File: py-utils/src/utils/setup/openldap/postupgradecmd.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    except Exception as e:
+      raise OpenldapPROVError(f'exception: {e}\n')
+
+  def process(self):
+    """Main processing function."""
+    try:
+      configFile = "/opt/seagate/cortx/utils/conf/openldap_config.yaml"
+      oldSampleFile = os.path.join(self.utils_tmp_dir, "openldap_config.yaml.sample.old")
+      newSampleFile = "/opt/seagate/cortx/utils/conf/openldap_config.yaml.sample"
+      unsafeAttributesFile = "/opt/seagate/cortx/utils/conf/openldap_config_unsafe_attributes.yaml"
```
Comment (shalakadharap) [inline py-utils/src/utils/setup/openldap/postupgradecmd.py:48]: Fetch it from Conf.get(index= 'util_config_file_index', key='install_path')
then append rest.
Follow same everywhere

File: py-utils/src/utils/setup/openldap/postupgradecmd.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+    for every parameter, check
+    - if it is marked as 'unsafe' in attributes file, skip
+    - if it marked as 'safe' in the attributes file
+        - diff the value in config and sample.old - if it is changed, skip
+        - if it is not changed,  we will overwrite the value in cfg file from sample.new
+        - if it does not exist in cfg file add the value from sample.new file to cfg file
+    - All the arrays in yaml are always overwritten"""
+    #If config file is not present then abort merging.
+    if not os.path.isfile(configFile):
+        Log.error(f'config file {configFile} does not exist')
```
Comment (shahidSTX) [inline py-utils/src/utils/setup/openldap/postupgradecmd.py:77]: General recommendation given by Foundation team is that avoid `f''` formatting. Please refer to other examples and use `%s` instead.

File: py-utils/src/utils/setup/openldap/postupgradecmd.py [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+  def merge_config(self, configFile:str, oldSampleFile:str, newSampleFile:str, unsafeAttributesFile:str, filetype:str):
+    """Core logic for updating config files during upgrade using conf store.
+    Following is algorithm from merge:
+    Iterate over all parameters sample.new file
+    for every parameter, check
+    - if it is marked as 'unsafe' in attributes file, skip
+    - if it marked as 'safe' in the attributes file
+        - diff the value in config and sample.old - if it is changed, skip
+        - if it is not changed,  we will overwrite the value in cfg file from sample.new
+        - if it does not exist in cfg file add the value from sample.new file to cfg file
```
Comment (shahidSTX) [inline py-utils/src/utils/setup/openldap/postupgradecmd.py:73]: Check char limit and set it to less than 80.

File: py-utils/src/utils/setup/openldap/openldap.py (hunk 1 above)
Comment (vimalraghwani20) [reply py-utils/src/utils/setup/openldap/openldap.py:60]: for preupgrade and postupgrade we dont pass the url. 
above handling will work for reset and cleanup too if we remove confi_url from setup command

File: py-utils/src/utils/setup/openldap/openldap.py (hunk 2 above)
Comment (vimalraghwani20) [reply py-utils/src/utils/setup/openldap/openldap.py]: fixed

File: py-utils/src/utils/setup/openldap/openldap.py (hunk 3 above)
Comment (vimalraghwani20) [reply py-utils/src/utils/setup/openldap/openldap.py]: fixed. 
config was also missing. moved to top of the file

File: py-utils/src/utils/setup/openldap/preupgradecmd.py (hunk 4 above)
Comment (vimalraghwani20) [reply py-utils/src/utils/setup/openldap/preupgradecmd.py]: fixed

File: py-utils/src/utils/setup/openldap/openldapsetup_prereqs.json (hunk 6 above)
Comment (vimalraghwani20) [reply py-utils/src/utils/setup/openldap/openldapsetup_prereqs.json:40]: fixed

File: py-utils/src/utils/setup/openldap/postupgradecmd.py (hunk 7 above)
Comment (vimalraghwani20) [reply py-utils/src/utils/setup/openldap/postupgradecmd.py]: fixed

File: py-utils/src/utils/setup/openldap/postupgradecmd.py (hunk 8 above)
Comment (vimalraghwani20) [reply py-utils/src/utils/setup/openldap/postupgradecmd.py]: fixed

**Review by shalakadharap — APPROVED** [review]

File: py-utils/src/utils/setup/openldap/postupgradecmd.py [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+from cortx.utils.conf_store import Conf
+from setupcmd import SetupCmd, OpenldapPROVError
+
+
+class PostUpgradeCmd(SetupCmd):
+
+  """Post Upgrade Setup Cmd."""
+
+  name = "postupgrade"
+  Log.init('OpenldapProvisioning','/var/log/seagate/utils/openldap',level='DEBUG')
```
Comment (bkirunge7) [inline py-utils/src/utils/setup/openldap/postupgradecmd.py:32]: Dont hardcode path, fetch from config

Comment (vimalraghwani20) [reply py-utils/src/utils/setup/openldap/postupgradecmd.py:32]: I am not sure about this path is coming from config or not. 
from config we get only /opt/seagate which is installation path not log path. 
@shalakadharap  @shahidSTX  can you please confirm. 

Note: I have use Log.init as is referring to other phases.

File: py-utils/src/utils/setup/openldap/openldap_setup.py (hunk 5 above)
Comment (vimalraghwani20) [reply py-utils/src/utils/setup/openldap/openldap_setup.py]: fixed

File: py-utils/src/utils/setup/openldap/postupgradecmd.py (hunk 10 above)
Comment (vimalraghwani20) [reply py-utils/src/utils/setup/openldap/postupgradecmd.py:32]: I have checked the content of cortx config file.
It does not have log file path. 
so we have to hardcode it. no choice 

[root@ssc-vm-5473 rv]# cat /etc/cortx/cortx.conf
{
  "install_path": "/opt/seagate",
  "health_provider": {
    "storage": "cortx/utils/discovery/mocked_health_gen",
    "server": "cortx/utils/discovery/mocked_health_gen"
  },
  "resource_map": {
    "location": "/var/cortx/resources/",
    "store_type": "json",
    "requests_file": "/var/cortx/resources/requests.json",
    "expiry_sec": 1800
  }
}

**Review by shahidSTX — APPROVED** [review]

**Pull request merged**